File: app/src/main/java/com/kira/kdownloader/settings/store/SharedPreferencesKeyValueStore.java
package com.kira.kdownloader.settings.store;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.lifecycle.LiveData;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public final class SharedPreferencesKeyValueStore implements KeyValueStore {
    public static final String PREFERENCES_NAME = "app_settings";
    private final SharedPreferences preferences;
    private final LiveData<String> changes;

    public SharedPreferencesKeyValueStore(Context context) { this(context, PREFERENCES_NAME); }

    public SharedPreferencesKeyValueStore(Context context, String name) {
        preferences = context.getApplicationContext().getSharedPreferences(name, Context.MODE_PRIVATE);
        changes = new PreferenceChanges(preferences);
    }

    @Override public String getString(String key, String defaultValue) {
        try {
            String value = preferences.getString(key, defaultValue);
            return value == null ? defaultValue : value;
        } catch (ClassCastException ignored) {
            return defaultValue;
        }
    }

    @Override public int getInt(String key, int defaultValue) {
        try {
            Object value = preferences.getAll().get(key);
            if (value instanceof Integer) return (Integer) value;
            if (value instanceof Long) return ((Long) value).intValue();
            if (value instanceof String) return Integer.parseInt((String) value);
        } catch (RuntimeException ignored) {}
        return defaultValue;
    }

    @Override public boolean getBoolean(String key, boolean defaultValue) {
        try {
            Object value = preferences.getAll().get(key);
            if (value instanceof Boolean) return (Boolean) value;
            if ("true".equals(value)) return true;
            if ("false".equals(value)) return false;
        } catch (RuntimeException ignored) {}
        return defaultValue;
    }

    @Override public boolean contains(String key) { return preferences.contains(key); }
    @Override public Set<String> keys() { return new HashSet<>(preferences.getAll().keySet()); }
    @Override public LiveData<String> getChanges() { return changes; }

    @Override public void edit(EditorAction action) {
        SharedPreferences.Editor editor = preferences.edit();
        action.apply(new KeyValueStore.Editor() {
            @Override public KeyValueStore.Editor putString(String key, String value) { editor.putString(key, value); return this; }
            @Override public KeyValueStore.Editor putInt(String key, int value) { editor.putInt(key, value); return this; }
            @Override public KeyValueStore.Editor putBoolean(String key, boolean value) { editor.putBoolean(key, value); return this; }
            @Override public KeyValueStore.Editor remove(String key) { editor.remove(key); return this; }
            @Override public KeyValueStore.Editor clear() { editor.clear(); return this; }
        });
        editor.apply();
    }

    private static final class PreferenceChanges extends LiveData<String>
            implements SharedPreferences.OnSharedPreferenceChangeListener {
        private final SharedPreferences preferences;

        PreferenceChanges(SharedPreferences preferences) { this.preferences = preferences; }
        @Override protected void onActive() { preferences.registerOnSharedPreferenceChangeListener(this); }
        @Override protected void onInactive() { preferences.unregisterOnSharedPreferenceChangeListener(this); }
        @Override public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) { setValue(key); }
    }
}

File: app/src/main/java/com/kira/kdownloader/settings/store/SharedPreferencesKeyValueStore.kt
package com.kira.kdownloader.settings.store

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow

/**
 * [KeyValueStore] backed by a named [SharedPreferences] file.
 *
 * SharedPreferences already survives app restarts and app updates (Section 1) and is included in
 * auto-backup. Writes use [SharedPreferences.Editor.apply], which is atomic and durable — a process
 * killed mid-write either sees the old value or the new one, never a corrupt one (Section 14).
 *
 * Reads are defensive: a value stored under an unexpected type returns the caller's default rather
 * than throwing [ClassCastException].
 */
class SharedPreferencesKeyValueStore(
    context: Context,
    name: String = PREFERENCES_NAME,
) : KeyValueStore {
    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(name, Context.MODE_PRIVATE)

    override fun getString(key: String, default: String): String =
        runCatching { prefs.getString(key, default) ?: default }.getOrDefault(default)

    override fun getInt(key: String, default: Int): Int = runCatching {
        when (val value = prefs.all[key]) {
            is Int -> value
            is Long -> value.toInt()
            is String -> value.toIntOrNull() ?: default
            else -> default
        }
    }.getOrDefault(default)

    override fun getBoolean(key: String, default: Boolean): Boolean = runCatching {
        when (val value = prefs.all[key]) {
            is Boolean -> value
            is String -> value.toBooleanStrictOrNull() ?: default
            else -> default
        }
    }.getOrDefault(default)

    override fun contains(key: String): Boolean = prefs.contains(key)

    override fun keys(): Set<String> = prefs.all.keys.toSet()

    override val changes: Flow<String?> = callbackFlow {
        val listener = SharedPreferences.OnSharedPreferenceChangeListener { _, key -> trySend(key) }
        prefs.registerOnSharedPreferenceChangeListener(listener)
        awaitClose { prefs.unregisterOnSharedPreferenceChangeListener(listener) }
    }

    override fun edit(block: KeyValueStore.Editor.() -> Unit) {
        val editor = prefs.edit()
        object : KeyValueStore.Editor {
            override fun putString(key: String, value: String) = apply { editor.putString(key, value) }
            override fun putInt(key: String, value: Int) = apply { editor.putInt(key, value) }
            override fun putBoolean(key: String, value: Boolean) = apply { editor.putBoolean(key, value) }
            override fun remove(key: String) = apply { editor.remove(key) }
            override fun clear() = apply { editor.clear() }
        }.block()
        editor.apply()
    }

    companion object {
        const val PREFERENCES_NAME = "app_settings"
    }
}
